### PR TITLE
add e2e case for upgrade tc version

### LINKED ISSUE #637: add e2e test for force upgrade
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

Add e2e test for force upgrade https://github.com/pingcap/tidb-operator/pull/631

### PR DESCRIPTION
**Pull request by dragonly**
### What problem does this PR solve?

Add e2e cases for TidbCluster upgrading and Fix #637 

### What is changed and how does it work?

Implemented *Upgrade TidbCluster* in https://docs.google.com/document/d/1wRw2SMs66MthjuUUGA5qxzpYw8LyglyA3p6fm-crItU/edit#heading=h.kggsbl7onab

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [x] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [x] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
None
```

### CODE REVIEW

File: tests/actions.go [hunk 1]
```diff
@@ -39,7 +39,7 @@ import (
 	pingcapErrors "github.com/pingcap/errors"
 	"github.com/pingcap/kvproto/pkg/metapb"
 	"github.com/pingcap/tidb-operator/pkg/apis/pingcap/v1alpha1"
-	"github.com/pingcap/tidb-operator/pkg/client/clientset/versioned"
+	pcCli "github.com/pingcap/tidb-operator/pkg/client/clientset/versioned"
```
Comment (handlerww) [inline tests/actions.go:42]: Will it be greater if we spilt the name change from this PR? These changes seem not too related to the "upgrade tc version"

Comment (dragonly) [reply tests/actions.go:42]: yes, this could be in a separate. Let's see if anyone else thinks it deserves another PR, and make a change then?

Comment (handlerww) [reply tests/actions.go:42]: It's better to cherry-pick, and easier to support version management methods in this period. Some users can just pick-up this PR to their branch on-demand.

File: tests/actions.go [hunk 2]
```diff
@@ -1511,78 +1516,74 @@ func (oa *operatorActions) pdMembersReadyFn(tc *v1alpha1.TidbCluster) (bool, err
...
+			log.Logf("TidbCluster: %q pd member(%s/%s) is not health", tcID, member.ID, member.Name)
 			return false, nil
 		}
 	}
 
-	pdServiceName := controller.PDMemberName(tcName)
-	pdPeerServiceName := controller.PDPeerMemberName(tcName)
-	if _, err := oa.kubeCli.CoreV1().Services(ns).Get(pdServiceName, metav1.GetOptions{}); err != nil {
-		log.Logf("failed to get service: %s/%s", ns, pdServiceName)
+	if _, err := oa.kubeCli.CoreV1().Services(ns).Get(pdSetName, metav1.GetOptions{}); err != nil {
```
Comment (DanielZhangQD) [inline tests/actions.go:1559]: It's confusing to use the set name here even if it has the same value as `controller.PDMemberName(tcName)`.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 3]
```diff
@@ -1558,8 +1559,313 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
 		oa.UpgradeTidbClusterOrDie(&tcCfg)
 		oa.CheckTidbClusterStatusOrDie(&tcCfg)
 	})
+
+	// test cases for tc upgrade
+	ginkgo.Context("upgrade should work correctly", func() {
+		ginkgo.It("for tc and components version", func() {
+			ginkgo.By("Deploy initial tc")
+			tc := fixture.GetTidbCluster(ns, "upgrade-version", utilimage.TiDBV4Prev)
+			pvRetain := corev1.PersistentVolumeReclaimRetain
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1568]: Is it necessary to use `Retain` here?

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 4]
```diff
@@ -1558,8 +1559,313 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+			tc.Spec.PVReclaimPolicy = &pvRetain
+			tc.Spec.PD.StorageClassName = pointer.StringPtr("local-storage")
+			tc.Spec.TiKV.StorageClassName = pointer.StringPtr("local-storage")
+			tc.Spec.TiDB.StorageClassName = pointer.StringPtr("local-storage")
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1572]: No need to set this explicitly, just use the default sc.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 5]
```diff
@@ -1558,8 +1559,313 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+			utiltc.MustCreateTCWithComponentsReady(genericCli, oa, tc, 5*time.Minute, 10*time.Second)
+
+			ginkgo.By("Update components configuration")
+			err := controller.GuaranteedUpdate(genericCli, tc, func() error {
+				pdCfg := v1alpha1.NewPDConfig()
+				tikvCfg := v1alpha1.NewTiKVConfig()
+				tidbCfg := v1alpha1.NewTiDBConfig()
+				pdCfg.Set("level", "info")
+				tikvCfg.Set("level", "info")
+				tidbCfg.Set("level", "info")
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1629]: Can we set some other configurations? The log level is configured by default in `fixture.GetTidbCluster`.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 6]
```diff
@@ -1558,8 +1559,313 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+					if op == "in" {
+						framework.ExpectEqual(int(tc.Spec.PD.Replicas), 3)
+					} else {
+						framework.ExpectEqual(int(tc.Spec.PD.Replicas), 5)
+					}
+				})
+			}
+		})
+
+		// similar to PD scale-in/scale-out case above, need to check no evict leader scheduler left
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1735]: The evict leader scheduler is not checked.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 7]
```diff
@@ -1558,8 +1559,313 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+		ginkgo.It("with bad PD config, then recover after force upgrading PD", func() {
+			ginkgo.By("Deploy initial tc with incorrect PD image")
+			tc := fixture.GetTidbCluster(ns, "force-upgrade-pd", utilimage.TiDBV4Prev)
+			tc.Spec.PD.BaseImage = "wrong-pd-image"
+			err := genericCli.Create(context.TODO(), tc)
+			framework.ExpectNoError(err, "failed to create TidbCluster %s/%s", ns, tc.Name)
+
+			ginkgo.By("Wait for 1 min and ensure no PD Pod exist")
+			err = wait.PollImmediate(5*time.Second, 1*time.Minute, func() (bool, error) {
+				_, err := c.CoreV1().Pods(ns).Get(tc.Name, metav1.GetOptions{})
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1806]: Why get `tc.Name`?

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 8]
```diff
@@ -1558,8 +1559,313 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+			ginkgo.By("Update PD Pod to correct image")
+			err = controller.GuaranteedUpdate(genericCli, tc, func() error {
+				tc.Spec.PD.BaseImage = "pingcap/pd"
+				return nil
+			})
+			framework.ExpectNoError(err, "failed to change PD Pod image")
+
+			ginkgo.By("Wait for 1 min and ensure no PD Pod exist")
+			err = wait.PollImmediate(5*time.Second, 1*time.Minute, func() (bool, error) {
+				_, err := c.CoreV1().Pods(ns).Get(tc.Name, metav1.GetOptions{})
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1824]: ditto

File: tests/actions.go (hunk 2 above)
Comment (dragonly) [reply tests/actions.go:1559]: fixed in next commits

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 3 above)
Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go:1568]: the detailed e2e test doc specifies "set PV reclaim policy to Retain"

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 4 above)
Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go:1572]: the detailed e2e test doc specifies "use local PV", maybe change to "use standard sc"?

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 5 above)
Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go:1629]: fixed in next commits

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 7 above)
Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go:1806]: fixed in next commits.

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 8 above)
Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go:1824]: fixed in next commits.

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 6 above)
Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go:1735]: fixed in next commits.

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 3 above)
Comment (DanielZhangQD) [reply tests/e2e/tidbcluster/tidbcluster.go:1568]: You're right, this case also covers the `Retain` and local PV scenarios, sorry that I forgot this.

File: tests/e2e/tidbcluster/tidbcluster.go (hunk 4 above)
Comment (DanielZhangQD) [reply tests/e2e/tidbcluster/tidbcluster.go:1572]: You're right, this case also covers the `Retain` and local PV scenarios, sorry that I forgot this.

**Comment by DanielZhangQD** [comment]
@dragonly Please fix the conflict.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 9]
```diff
@@ -1557,8 +1558,327 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+			ginkgo.By("Check components configuration")
+			pdMemberName := controller.PDMemberName(tc.Name)
+			pdSts, err := stsGetter.StatefulSets(ns).Get(pdMemberName, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to get StatefulSet %s/%s", ns, pdMemberName)
+			pdCmName := member.FindConfigMapVolume(&pdSts.Spec.Template.Spec, func(name string) bool {
+				return strings.HasPrefix(name, controller.PDMemberName(tc.Name))
+			})
+			pdCm, err := c.CoreV1().ConfigMaps(ns).Get(pdCmName, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to get ConfigMap %s/%s", ns, pdCm)
+			gomega.Expect(pdCm.Data["config-file"]).To(gomega.ContainSubstring("level = \"info\""))
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1647]: This needs to match with the above config?

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 10]
```diff
@@ -1557,8 +1558,327 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+
+			tikvMemberName := controller.TiKVMemberName(tc.Name)
+			tikvSts, err := stsGetter.StatefulSets(ns).Get(tikvMemberName, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to get StatefulSet %s/%s", ns, tikvMemberName)
+			tikvCmName := member.FindConfigMapVolume(&tikvSts.Spec.Template.Spec, func(name string) bool {
+				return strings.HasPrefix(name, controller.TiKVMemberName(tc.Name))
+			})
+			tikvCm, err := c.CoreV1().ConfigMaps(ns).Get(tikvCmName, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to get ConfigMap %s/%s", ns, tikvCmName)
+			gomega.Expect(tikvCm.Data["config-file"]).To(gomega.ContainSubstring("level = \"info\""))
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1657]: ditto

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 11]
```diff
@@ -1557,8 +1558,327 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+
+			tidbMemberName := controller.TiDBMemberName(tc.Name)
+			tidbSts, err := stsGetter.StatefulSets(ns).Get(tidbMemberName, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to get StatefulSet %s/%s", ns, tidbMemberName)
+			tidbCmName := member.FindConfigMapVolume(&tidbSts.Spec.Template.Spec, func(name string) bool {
+				return strings.HasPrefix(name, controller.TiDBMemberName(tc.Name))
+			})
+			tidbCm, err := c.CoreV1().ConfigMaps(ns).Get(tidbCmName, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to get ConfigMap %s/%s", ns, tidbCmName)
+			gomega.Expect(tidbCm.Data["config-file"]).To(gomega.ContainSubstring("level = \"info\""))
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1667]: ditto

**Comment by dragonly** [comment]
The failing case "TiDBCluster upgrade should work correctly while concurrently scale PD out" is strange.
When running in e2e, it will stuck where pd is first upgraded and immediately scaled out. The pd-4 will complain about unmounted volume named "pd", and the PVC for volume "pd" is waiting for PV to be available, and the local-path-provisioner is sitting there doing nothing.
When running locally with manual operation, things are good.

File: pkg/manager/member/tikv_upgrader.go [hunk 12]
```diff
@@ -264,6 +264,7 @@ func endEvictLeaderbyStoreID(deps *controller.Dependencies, tc *v1alpha1.TidbClu
 	}
 	var err error
 
+	klog.Infof("tikv: start calling EndEvictLeader on PD")
```
Comment (DanielZhangQD) [inline pkg/manager/member/tikv_upgrader.go:267]:
```suggestion
```

**Review by DanielZhangQD — APPROVED** [review]
LGTM

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 13]
```diff
@@ -1574,6 +1573,366 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+			tikvCmName := member.FindConfigMapVolume(&tikvSts.Spec.Template.Spec, func(name string) bool {
+				return strings.HasPrefix(name, controller.TiKVMemberName(tc.Name))
+			})
+			tikvCm, err := c.CoreV1().ConfigMaps(ns).Get(tikvCmName, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to get ConfigMap %s/%s", ns, tikvCmName)
+			log.Logf("TiKV config:\n%s", tikvCm.Data["config-file"])
+			gomega.Expect(tikvCm.Data["config-file"]).To(gomega.ContainSubstring("status-thread-pool-size = 1"))
+
+			ginkgo.By("Wait for TiDB to be in UpgradePhase")
+			utiltc.MustWaitForTiDBPhase(cli, tc, v1alpha1.UpgradePhase, 3*time.Minute, 10*time.Second)
```
Comment (csuzhangxc) [inline tests/e2e/tidbcluster/tidbcluster.go:1691]: Can we always observe these intermediate phases?

Comment (dragonly) [reply tests/e2e/tidbcluster/tidbcluster.go:1691]: @csuzhangxc I think we can as long as the interval is small enough.
Do you have any better solution of waiting for the correct condition?

Comment (csuzhangxc) [reply tests/e2e/tidbcluster/tidbcluster.go:1691]: OK. No better solution now.

**Review by csuzhangxc — APPROVED** [review]
LGTM

**Pull request merged**